ReactNativeClient/lib/joplin-renderer/MdToHtml/rules/embedvideo.js

```diff
@@ -0,0 +1,109 @@
+'use strict';
+
+// The youtubeParser is from http://stackoverflow.com/a/8260383
+function youtubeParser(url) {
+  var regExp = /^.*((youtu.be\/)|(v\/)|(\/u\/\w\/)|(embed\/)|(watch\?))\??v?=?([^#&?]*).*/;
+  var match = url.match(regExp);
+  if (match && match[7].length == 11) {
+    return match[7];
+  } else {
+    return url;
+  }
+}
+
+function videoEmbed(md) {
+  function videoReturn(state, silent) {
+    var serviceEnd,
+      serviceStart,
+      token,
+      videoID,
+      theState = state;
+    const oldPos = state.pos;
+
+    var extractURL = /\[(youtube)\]\([\s]*(.*?)[\s]*[)]/im;
+
+    if (state.src.charCodeAt(state.pos) !== 0x5b /* [ */) {
+      return false;
+    }
+
+    const match = extractURL.exec(state.src.slice(state.pos, state.src.length));
+
+    if (!match) {
+      return false;
+    }
+
+    if (match.length < 3) {
+      return false;
+    }
+
+    const service = match[1];
+    videoID = match[2];
+    if (service.toLowerCase() == 'youtube') {
+      videoID = youtubeParser(videoID);
+    } else {
+      return false;
+    }
+
+    serviceStart = oldPos + 1;
+    serviceEnd = md.helpers.parseLinkLabel(state, state.pos, false);
+
+    //
+    // We found the end of the link, and know for a fact it's a valid link;
+    // so all that's left to do is to call tokenizer.
+    //
+    if (!silent) {
+      theState.pos = serviceStart;
+      theState.service = state.src.slice(serviceStart, serviceEnd);
+      // constructor
+      const newState = new state.md.inline.State(
+        service,
+        theState.md,
+        theState.env,
+        [] // tokens=[]
+      );
+
+      // console.log(newState);
+      newState.md.inline.tokenize(newState);
+
+      token = theState.push('video', '');
+      token.videoID = videoID;
+      token.service = service;
+      token.url = match[2];
+      token.level = theState.level;
+    }
+
+    theState.pos += theState.src.indexOf(')', theState.pos);
+    return true;
+  }
+
+  return videoReturn;
+}
+
+function tokenizeYoutube(videoID) {
+  var embedStart =
+    '<div><iframe  width="560" height="315" src="https://www.youtube.com/embed/';
+  var embedEnd =
+    '"  frameborder="0" allowfullscreen allow="accelerometer; autoplay; encrypted-media; gyroscope; picture-in-picture"></iframe></div>';
+
+  return embedStart + videoID + embedEnd;
+}
```

Comment:
"tokenize" means converting a string to a stream of tokens. That's not what this function does.

Reply:
I'm writing a post that explains everything, I'll be done with it in an hour and I'll update the description of the PR for better understanding how everything works since this rule is heavily based from the repo I mentioned, I also have some gaps in my knowledge about this topic, but I'll try to explain it most of what I understand(after 2 days of research in the same)